### DISCUSSION TITLE
Exporting a trained neural network as an FMU

### DISCUSSION THREAD

**Discussion by SriRama29** [ThummeTo/FMIExport.jl#43] [General]
Hi,
Let us assume I have the parameters of a  neural network that I have trained using the neural ODE framework. Is it possible for me to use this package to export it as an FMU?

I have spent some time trying to understand the examples. I think the NeuralFMU example is more complicated than what I need. I think I might need to make my code from a mix of the bouncing ball example and the neural FMU example. 

I have not yet figured out how I would load the parameters into the code and export it as an FMU. (Any advice is appreciated!)

But before I attempt to work further on it, I wanted to verify if I can even accomplish my requirement with the package 😄 

Looking forward to your response!

**Comment by ThummeTo** [comment]
Hi,

**short answer:** 
you can use this lib to export your neural ODE as FMU!

**longer answer:**
As proposed, I would start with the bouncing ball example. A neural ODE is nothing but a neural network that is interpreted as right-hand side of an ODE, that gets solved by a numerical ODE solver. So to export this as FMU, you just need to define your function `f` that maps state `x` to state derivative `dx`, this is just your NN inference. For the bouncing ball its the second order ODE for a mass that is accelerated by gravity (so the most simple system existing). The more advanced things for discontinuous systems (event indicators, condition and affect) you can skip for your (I think continuous) neural ODE. So to conclude, replacing the `f` and setting the remaining functions to neutral should do the trick. 

Best regards!

Remark: We should make a tutorial for this, because this is not the first time I get ask this :-)

**Reply by ThummeTo** [reply]
I moved this to discussions :-)

**Reply by SriRama29** [reply]
Thanks for the response! Yeah, I tried something like that and after facing some problems during compilation, I think I have a version of the code that should work. I have not gotten the time to test it out yet, so I will keep you updated. 

I defined the neural network using the Chain and Dense type you had used in the neural FMU example. I also hard coded the parameters while defining the network using loops. Only point of concern for me is how to define the activation functions. In my case, I have an elu and a sigmoid function. In my case, I just defined the functions by myself and used it as input when defining my network. 

I will let you know if it works well! thanks.

**Reply by SriRama29** [reply]
Also, a tutorial for this would be very useful! :D

**Comment by SriRama29** [comment]
Hello! So I managed to get my code to compile and to get an fmu from the program. But somehow I am finding it hard to get the fmu to run in my target program. 

So I decided to create an fmu of the bouncing ball problem and try to run it in my program, but I keep getting an error that says,

InitError(mod=:LibCURL_jll, error=ErrorException("could not load library "libcurl-4.dll")

from my program. The simulation tool into which I am importing my fmu is called Cruise M. 

My question is, is this a problem with the way I have exported the bouncing ball example? Is there maybe an issue with my target system (missing software or something like that?)

Any help would be greatly appreciated!

**Reply by ThummeTo** [reply]
Do you have access to another FMU simulator (many modeling tools support for import and simulation of FMUs via drag-and-drop). It is always a good idea to check with another simulator first. If you don't have another FMI-importing tool, you can check with [FMPy](https://github.com/CATIA-Systems/FMPy) (if you are familiar with Python).

However this error message sounds like a DLL was not shipped with the FMU, but is required ... or the DLL is expected to be part of your OS and is not there ...

**Reply by SriRama29** [reply]
Hi! so i tried to load the bouncing ball fmu using FMPy. I get this error from it, 

Failed to validate modelDescription.xml:
- The variable "sticking" (line 22) is of type Integer and must have variability != "continuous".

**Reply by ThummeTo** [reply]
is this still the case with the current releases?
